src/v/cloud_storage/topic_manifest_downloader.cc

```diff
@@ -0,0 +1,110 @@
+
+#include "base/outcome.h"
+#include "cloud_storage/topic_manifest.h"
+#include "cloud_storage/topic_path_utils.h"
+#include "cloud_storage/types.h"
+#include "container/fragmented_vector.h"
+#include "hashing/xx.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/lowres_clock.hh>
+
+namespace cloud_storage {
+
+topic_manifest_downloader::topic_manifest_downloader(
+  const cloud_storage_clients::bucket_name bucket,
+  std::optional<ss::sstring> hint,
+  const model::topic_namespace topic,
+  remote& remote)
+  : bucket_(bucket)
+  , label_hint_(std::move(hint))
+  , topic_(topic)
+  , remote_(remote) {}
+
+ss::future<result<find_topic_manifest_outcome, error_outcome>>
+topic_manifest_downloader::download_manifest(
+  retry_chain_node& parent_retry,
+  ss::lowres_clock::time_point deadline,
+  model::timestamp_clock::duration backoff,
+  topic_manifest* manifest) {
+    retry_chain_node retry_node(deadline, backoff, &parent_retry);
+
+    // First look for topic manifests with a label.
+    ss::sstring remote_label_str{""};
+    if (label_hint_.has_value()) {
+        remote_label_str = *label_hint_;
+    }
+    const auto labeled_manifest_filter = fmt::format(
+      "{}/{}", labeled_topic_manifest_root(topic_), remote_label_str);
+    auto list_res = co_await remote_.list_objects(
+      bucket_,
+      retry_node,
+      cloud_storage_clients::object_key{labeled_manifest_filter});
+    if (list_res.has_error()) {
+        co_return error_outcome::manifest_download_error;
+    }
+    // If there's more than one, callers will need to pass a label (or a more
+    // specific one).
+    chunked_vector<ss::sstring> list_contents;
+    for (auto& entry : list_res.value().contents) {
+        if (entry.key.ends_with("topic_manifest.bin")) {
+            list_contents.emplace_back(entry.key);
+        }
+    }
+    if (list_contents.size() > 1) {
+        co_return find_topic_manifest_outcome::multiple_matching_manifests;
+    }
+    // If there's exactly one, presume it's the one we care about. Since
+    // labeled manifests are newer, they take precedence over prefixed
+    // manifests.
+    if (list_contents.size() == 1) {
+        const remote_manifest_path labeled_manifest(list_contents[0]);
+        auto manifest_res = co_await remote_.download_manifest_bin(
+          bucket_, labeled_manifest, *manifest, retry_node);
+        if (manifest_res == cloud_storage::download_result::success) {
+            co_return find_topic_manifest_outcome::success;
+        }
+        // Regardless of the outcome (i.e. even not-found), return an error. If
+        // we had a list result but it was deleted, something is suspicious, so
+        // don't proceed.
+        co_return error_outcome::manifest_download_error;
+    }
+
+    // Then look for prefixed binary manifests. If we find one, return it:
+    // since they're newer than JSON manifests, they take precedence.
+    const remote_manifest_path prefixed_bin_path(
```

Comment:
Is it worth it to encapsulate repeated logic of attempting `remote_.download_manifest_bin()` in its own function?

Reply:
I don't think so. It's a single-line call. Plus they handle results differently.